src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs

```diff
@@ -169,6 +174,51 @@ protected static ValueSource<TextAndVersion> CreateRecoverableText(TextLoader lo
             }
         }
 
+        protected static TextAndVersion LoadTextSync(TextLoader loader, DocumentId documentId, SolutionServices services, bool reportInvalidDataException, CancellationToken cancellationToken)
+        {
+            int retries = 0;
+
+            while (true)
+            {
+                try
+                {
+                    using (ExceptionHelpers.SuppressFailFast())
+                    {
+                        var result = loader.LoadTextAndVersion(services.Workspace, documentId, cancellationToken);
+                        return result;
+                    }
+                }
+                catch (OperationCanceledException)
+                {
+                    // if load text is failed due to a cancellation, make sure we propagate it out to the caller
+                    throw;
+                }
+                catch (IOException e)
+                {
+                    if (++retries > MaxRetries)
+                    {
+                        services.Workspace.OnWorkspaceFailed(new DocumentDiagnostic(WorkspaceDiagnosticKind.Failure, e.Message, documentId));
+                        return TextAndVersion.Create(SourceText.From(string.Empty, Encoding.UTF8), VersionStamp.Default, documentId.GetDebuggerDisplay());
+                    }
+
+                    // fall out to try again
+                }
+                catch (InvalidDataException e)
+                {
+                    // TODO: Adjust this behavior in the future if we add support for non-text additional files
+                    if (reportInvalidDataException)
+                    {
+                        services.Workspace.OnWorkspaceFailed(new DocumentDiagnostic(WorkspaceDiagnosticKind.Failure, e.Message, documentId));
+                    }
+
+                    return TextAndVersion.Create(SourceText.From(string.Empty, Encoding.UTF8), VersionStamp.Default, documentId.GetDebuggerDisplay());
+                }
+
+                // try again after a delay
+                Task.Delay(RetryDelay).Wait(cancellationToken);
```

Comment:
I feel like we should add something to PortableShim to allow it - it is actually there on all the platforms we really run on I think.  This will not help anything, as we'll still end up chewing up the thread with the Wait, and so the continuation may not get to run.